plugins/pixart-rf/fu-pxi-device.c

```diff
@@ -0,0 +1,515 @@
+static gboolean
+fu_pxi_device_set_feature (FuPxiDevice *self, const guint8 *data, guint datasz, GError **error)
+{
+#ifdef HAVE_HIDRAW_H
+	if (g_getenv ("FWUPD_PIXART_RF_VERBOSE") != NULL)
+		fu_common_dump_raw (G_LOG_DOMAIN, "SetFeature", data, datasz);
+	return fu_udev_device_ioctl (FU_UDEV_DEVICE (self),
+				     HIDIOCSFEATURE(datasz), (guint8 *) data,
+				     NULL, error);
+#else
+	g_set_error_literal (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_NOT_SUPPORTED,
+			     "<linux/hidraw.h> not available");
+	return FALSE;
+#endif
+}
+#endif
+
+static gboolean
+fu_pxi_device_get_feature (FuPxiDevice *self, guint8 *data, guint datasz, GError **error)
+{
+#ifdef HAVE_HIDRAW_H
+	if (!fu_udev_device_ioctl (FU_UDEV_DEVICE (self),
+				   HIDIOCGFEATURE(datasz), data,
+				   NULL, error)) {
+		return FALSE;
+	}
+	if (g_getenv ("FWUPD_PIXART_RF_VERBOSE") != NULL)
+		fu_common_dump_raw (G_LOG_DOMAIN, "GetFeature", data, datasz);
+	return TRUE;
+#else
+	g_set_error_literal (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_NOT_SUPPORTED,
+			     "<linux/hidraw.h> not available");
+	return FALSE;
+#endif
+}
+
+static guint16
+fu_pxi_device_calculate_checksum (const guint8 *data, gsize sz)
+{
+	guint16 checksum = 0;
+	for (gsize idx = 0; idx < sz; idx++)
+		checksum += (guint16) data[idx];
+	return checksum;
+}
+
+/* FIXME: is @port required? */
+static gboolean
+fu_pxi_device_wait_notify (FuPxiDevice *self,
+			   goffset port,
+			   guint8 *status,
+			   guint16 *checksum,
+			   GError **error)
+{
+	guint8 res[FU_PXI_DEVICE_OTA_BUF_SZ] = {
+		PXI_HID_DEV_OTA_INPUT_REPORT_ID,
+		0x0,
+	};
+	if (!fu_udev_device_pread_full (FU_UDEV_DEVICE (self),
+					port, res, sizeof(res) - port,
+					error))
+		return FALSE;
+	if (g_getenv ("FWUPD_PIXART_RF_VERBOSE") != NULL)
+		fu_common_dump_raw (G_LOG_DOMAIN, "notify", res, sizeof(res));
+	if (status != NULL) {
+		if (!fu_common_read_uint8_safe (res, sizeof(res), 0x01,
+						status, error))
+			return FALSE;
+	}
+	if (checksum != NULL) {
+		if (!fu_common_read_uint16_safe (res, sizeof(res), 0x02,
+						 checksum, G_LITTLE_ENDIAN, error))
+			return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_pxi_device_fw_object_create (FuPxiDevice *self, FuChunk *chk, GError **error)
+{
+	g_autoptr(GByteArray) req = g_byte_array_new ();
+	guint8 res[FU_PXI_DEVICE_OTA_BUF_SZ] = { PXI_HID_DEV_OTA_INPUT_REPORT_ID };
+
+	/* request */
+	fu_byte_array_append_uint8 (req, PXI_HID_DEV_OTA_OUTPUT_REPORT_ID);
+	fu_byte_array_append_uint8 (req, FU_PXI_DEVICE_CMD_FW_OBJECT_CREATE);
+	fu_byte_array_append_uint32 (req, chk->address, G_LITTLE_ENDIAN);
+	fu_byte_array_append_uint32 (req, chk->data_sz, G_LITTLE_ENDIAN);
+	if (!fu_udev_device_pwrite_full (FU_UDEV_DEVICE (self), 0x0,
+					 req->data, req->len, error))
+		return FALSE;
+
+	/* reply */
+	if (!fu_udev_device_pread_full (FU_UDEV_DEVICE (self), 0x0,
+					res, sizeof(res), error))
+		return FALSE;
+	g_usleep (30 * 1000);
+
+	/* res seems unused */
```

Comment:
@jyu99 is that expected?

Reply:
I will remove the delay thank your mention

Reply:
It might be the hardware *needs* a read after the write; this is fine, but it would be nice to document this if so.